### PR TITLE
Add Line Numbers to Dex Dump

### PR DESCRIPTION
**Pull request by alonalbert**
I rewrote the DexDump parsing into a dedicated class.

I changed the implementation to be (IMO) less brittle and more readable.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -30,13 +30,14 @@ kotlin {
...
                 implementation("org.jetbrains.jewel:jewel-int-ui-decorated-window-241:${extra["jewel.version"] as String}")
-                implementation("org.jetbrains.skiko:skiko-awt-runtime-macos-arm64:${extra["skiko.version"] as String}")
                 implementation("org.jetbrains.compose.components:components-splitpane-desktop:${extra["compose.version"] as String}")
                 implementation("com.fifesoft:rsyntaxtextarea:${extra["rsyntaxtextarea.version"] as String}")
                 implementation("com.fifesoft:rstaui:${extra["rstaui.version"] as String}")
                 implementation("net.java.dev.jna:jna:${extra["jna.version"] as String}")
                 implementation("androidx.collection:collection:${extra["collections.version"] as String}")
                 implementation("org.jetbrains.androidx.lifecycle:lifecycle-runtime:${extra["lifecycle.version"] as String}")
+                runtimeOnly("org.jetbrains.skiko:skiko-awt-runtime-linux-x64:${extra["skiko.version"] as String}")
+                runtimeOnly("org.jetbrains.skiko:skiko-awt-runtime-macos-arm64:${extra["skiko.version"] as String}")
```
Comment (romainguy) [inline build.gradle.kts:40]: This breaks on macOS, the native library can't be found if it's not `implementation()` :/

File: src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ */
+
+package dev.romainguy.kotlin.explorer.dex
+
+import dev.romainguy.kotlin.explorer.BuiltInKotlinClass
+import dev.romainguy.kotlin.explorer.consumeUntil
+
+private val PositionRegex = Regex("^\\s*0x(?<address>[0-9a-f]+) line=(?<line>\\d+)$")
+
+private const val CLASS_START = "Class #"
```
Comment (romainguy) [inline src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt:24]: I've been following the `SnakeCase` style for constants here.

File: src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+internal class DexDumpParser(text: String) {
+    private val lines = text.lineSequence().iterator()
+    fun parseDexDump(): String {
+        val classes = buildList {
+            while (lines.consumeUntil(CLASS_START)) {
+                add(readClass())
+            }
+        }
+        return classes
+            .filter { it.methods.isNotEmpty() }
```
Comment (romainguy) [inline src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt:39]: You can combine those two filters in a single one

File: src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    private fun readMethod(className: String): DexMethod {
+        val (name, type) = lines.next().substringAfterLast(".").split(':', limit = 2)
+        val instructions = readInstructions()
+        lines.consumeUntil(POSITIONS)
+        val positions = readPositions()
+        val code = buildString {
+            instructions.forEach {
+                val lineNumber = positions[it.address]
+                val prefix = when (lineNumber != null) {
```
Comment (romainguy) [inline src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt:66]: Use an if/else here instead of when/true/false

**Review by romainguy — CHANGES_REQUESTED** [review]
Thanks, I left a couple of comments. A few more things:

- There are now 2 empty lines between methods in the output instead of just 1. I remember I had to do something to deal with this in the original implementation as well.
- There seems to be a bug with naming classes. For instance I got this in a dump: `class IntOffset..ExternalSyntheticBackport0` (note the two dots)
- Could you add a menu to turn debug line numbers on/off? It would make it more obvious to users what those numbers are, but more importantly it would make copy/pasting out of the tool easier when line numbers are disabled.
- Your change breaks another feature: when clicking on a line with a jump (goto/if/etc.) in the DEX output, the tool will draw a line to show where the jump will go. If the target line has a debug line number in the output, the feature stops working (prob. because the regex I use assumes whitespace at the beginning; one of the reasons why I wanted to change the parsers to output a list of class/methods/instructions)

File: build.gradle.kts (hunk 1 above)
Comment (alonalbert) [reply build.gradle.kts:40]: Done

File: src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt (hunk 2 above)
Comment (alonalbert) [reply src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt:24]: Done

File: src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt (hunk 3 above)
Comment (alonalbert) [reply src/jvmMain/kotlin/dev/romainguy/kotlin/explorer/dex/DexDumpParser.kt:39]: Done

**Comment by romainguy** [comment]
Thanks for the updates! One last issue, there are still 2 empty lines between methods in the dump instead of 1.

**Review by romainguy — APPROVED** [review]
Awesome, thank you!

**Pull request merged**